examples_tests/61.BlitFilterTest/main.cpp

```diff
@@ -0,0 +1,1561 @@
+	core::smart_refctd_ptr<ICPUImage> image = ICPUImage::create(std::move(imageParams));
+	image->setBufferAndRegions(core::smart_refctd_ptr(imageBuffer), imageRegions);
+
+	if (fillWithTestData)
+	{
+		double pixelValueUpperBound = 20.0;
+		if (asset::isNormalizedFormat(format))
+			pixelValueUpperBound = 1.00000000001;
+
+		std::uniform_real_distribution<double> dist(0.0, pixelValueUpperBound);
+		std::mt19937 prng;
+
+		uint8_t* bytePtr = reinterpret_cast<uint8_t*>(image->getBuffer()->getPointer());
+		for (uint64_t k = 0u; k < dims[2]; ++k)
+		{
+			for (uint64_t j = 0u; j < dims[1]; ++j)
+			{
+				for (uint64_t i = 0; i < dims[0]; ++i)
+				{
+					double decodedPixel[4] = { 0 };
+					for (uint32_t ch = 0u; ch < asset::getFormatChannelCount(format); ++ch)
+						decodedPixel[ch] = dist(prng);
+
+					const uint64_t pixelIndex = (k * dims[1] * dims[0]) + (j * dims[0]) + i;
+					asset::encodePixelsRuntime(format, bytePtr + pixelIndex * asset::getTexelOrBlockBytesize(format), decodedPixel);
+				}
+			}
+		}
+	}
+
+	return image;
+}
+
+static inline asset::IImageView<asset::ICPUImage>::E_TYPE getImageViewTypeFromImageType_CPU(const asset::IImage::E_TYPE type)
+{
+	switch (type)
+	{
+	case asset::IImage::ET_1D:
+		return asset::ICPUImageView::ET_1D;
+	case asset::IImage::ET_2D:
+		return asset::ICPUImageView::ET_2D;
+	case asset::IImage::ET_3D:
+		return asset::ICPUImageView::ET_3D;
+	default:
+		__debugbreak();
+		return static_cast<asset::IImageView<asset::ICPUImage>::E_TYPE>(0u);
+	}
+}
+
+static inline video::IGPUImageView::E_TYPE getImageViewTypeFromImageType_GPU(const video::IGPUImage::E_TYPE type)
+{
+	switch (type)
+	{
+	case video::IGPUImage::ET_1D:
+		return video::IGPUImageView::ET_1D;
+	case video::IGPUImage::ET_2D:
+		return video::IGPUImageView::ET_2D;
+	case video::IGPUImage::ET_3D:
+		return video::IGPUImageView::ET_3D;
+	default:
+		__debugbreak();
+		return static_cast<video::IGPUImageView::E_TYPE>(0u);
+	}
+}
+
+class CBlitFilter
+{
+private:
+	struct alignas(16) vec3_aligned
+	{
+		float x, y, z;
+	};
+
+	struct alignas(16) uvec3_aligned
+	{
+		uint32_t x, y, z;
+	};
+
+public:
+	static constexpr uint32_t NBL_GLSL_DEFAULT_WORKGROUP_DIM = 16u;
+	static constexpr uint32_t NBL_GLSL_DEFAULT_WORKGROUP_SIZE = 256u;
+	static constexpr uint32_t NBL_GLSL_DEFAULT_BIN_COUNT = 256u;
+
+	struct alpha_test_push_constants_t
+	{
+		float referenceAlpha;
+	};
+
+	struct blit_push_constants_t
+	{
+		uvec3_aligned inDim;
+		uvec3_aligned outDim;
+		vec3_aligned negativeSupport;
+		vec3_aligned positiveSupport;
+		uvec3_aligned windowDim;
+		uvec3_aligned phaseCount;
+		uint32_t windowsPerWG;
+		uint32_t axisCount;
+	};
+
+	struct normalization_push_constants_t
+	{
+		uvec3_aligned outDim;
+		uint32_t inPixelCount;
+		float referenceAlpha;
+	};
+
+	struct dispatch_info_t
+	{
+		uint32_t wgCount[3];
+	};
+
+	CBlitFilter(video::ILogicalDevice* logicalDevice, const uint32_t smemSize = 16*1024u) : device(logicalDevice), sharedMemorySize(smemSize)
```

Comment:
@Erfan-Ahmadi does `logicalDevice` report the max smem size?

Reply:
also make a `create` static method, you want to be able to bail if something fails to create

Reply:
> @Erfan-Ahmadi does `logicalDevice` report the max smem size (via its physical device)?

No, I don't think so

Reply:
can you make it report it soon please?

Reply:
Yea sure, I'll put it in my to-do list

Reply:
But I think it needs vendor specific calls. I couldn't find any function that gives this info in Vulkan

Reply:
GL_MAX_COMPUTE_SHARED_MEMORY_SIZE this is for gl

Reply:
> But I think it needs vendor specific calls. I couldn't find any function that gives this info in Vulkan

its a device Limit `maxComputeSharedMemorySize`

Reply:
Ok cool then it's settled

Reply:
It's done @devshgraphicsprogramming `physdev.m_limits.maxComputeSharedMemorySize`

here is the commit: https://github.com/Devsh-Graphics-Programming/Nabla/commit/6ce6ad1a1cf5fc6166c9f24abbc765725582619a

 it's pushed to my branch for now in `erfan_todos`

Also I tested in my PC and both GL and Vulkan report the same correct number for `maxComputeSharedMemorySize`

Reply:
@achalpandeyy its ready.

Reply:
@achalpandeyy you just need to merge `master` and you have access to that

Reply:
Resolved.

Reply:
> also make a create static method, you want to be able to bail if something fails to create

Also, resolved.